torch/_prims/utils.py

```diff
@@ -0,0 +1,468 @@
+
+    raise ValueError("Invalid dtype!")
+
+
+_type_to_dtype_map = {
+    bool: torch.bool,
+    int: torch.int64,
+    float: torch.float64,
+    complex: torch.complex128,
+}
+
+
+def type_to_dtype(typ: type) -> torch.dtype:
+    """
+    Computes the corresponding dtype for a Number type.
+    """
+    return _type_to_dtype_map[typ]
+
+
+_ordered_types = (bool, int, float, complex)
+
+
+def get_higher_type(a: type, b: type) -> type:
+    """
+    Returns the higher of the two given Number types.
+
+    The types are ordered bool -> int -> float -> complex.
+    """
+    # Type checking
+    assert a in _ordered_types
+    assert b in _ordered_types
+
+    if a is b:
+        return a
+
+    for typ in _ordered_types:
+        if a is typ:
+            return b
+        if b is typ:
+            return a
+
+    raise ValueError("Unknown Python scalar type!")
+
+
+# Returns the higher of two torch datatypes a and b or, if the two
+#   are not ordered relative to each other, the next
+#   higher datatype
+def get_higher_dtype(
+    a: Union[torch.dtype, torch.Tensor, Number],
+    b: Union[torch.dtype, torch.Tensor, Number],
+) -> torch.dtype:
+    """
+    Computes the "lowest" datatype that is weakly
+    "higher" than both a and b.
+    """
+
+    # Type checking
+    assert isinstance(a, (torch.dtype, torch.Tensor, Number))
+    assert isinstance(b, (torch.dtype, torch.Tensor, Number))
+
+    def _extract_dtype(x):
+        if isinstance(x, torch.dtype):
+            return x
+        if isinstance(x, torch.Tensor):
+            return x.dtype
+        if isinstance(x, Number):
+            return type_to_dtype(type(x))
+
+        raise RuntimeError("Unexpected type given to _extract_dtype!")
+
+    a, b = _extract_dtype(a), _extract_dtype(b)
+
+    if a is b:
+        return a
+
+    ordered_datatypes = (
+        (torch.bool,),
+        (torch.uint8, torch.int8),
+        (torch.int16,),
+        (torch.int32,),
+        (torch.int64,),
+        (torch.float16, torch.bfloat16),
+        (torch.float32,),
+        (torch.float64,),
+        (torch.complex64,),
+        (torch.complex128,),
+    )
+
+    for idx, dtypes in enumerate(ordered_datatypes):
+        if a in dtypes and b in dtypes:
+            return ordered_datatypes[idx + 1][0]
+        if a in dtypes:
+            return b
+        if b in dtypes:
+            return a
+
+
+def is_lesser_type(a: type, b: type) -> bool:
+    """
+    Compares two types, a and b, returning True if a is "less" than b.
+
+    The comparison is determined by the following type ordering: bool, int, float, complex.
+    """
+    ordered_types = (
+        bool,
+        int,
+        float,
+        complex,
+    )
+
+    assert a in ordered_types
+    assert b in ordered_types
+
+    for typ in ordered_types:
+        if a == typ:
+            return True
+        if b == typ:
+            return False
+
+    raise RuntimeError("Unexpected termination!")
+
+
+def check_same_dtype(*args):
```

Comment:
It would be nice to recast these in a form that makes it clear that "combining" these dtypes commutative/associative/distributive; from the looping form here it is not obvious that those high level properties hold